app/lib/ecies/ecies.ts

```diff
@@ -0,0 +1,246 @@
+/**
+ * ECIES (Elliptic Curve Integrated Encryption Scheme) is a public-key encryption scheme that uses elliptic curve cryptography.
+ *
+ * This is an implementation of ECIES using the secp256k1 curve, ChaCha20-Poly1305 for encryption, and HKDF for key derivation.
+ *
+ * ## How it works:
+ * - **Asymmetric**: Encrypt with recipient's public key, decrypt with their private key
+ * - **Ephemeral keys**: Each message generates a new temporary key pair for forward secrecy
+ * - **Hybrid approach**: Uses ECDH to create shared secret, then symmetric encryption for efficiency
+ * - **Self-contained**: Ephemeral public key travels with the encrypted message
+ *
+ * The official specification is in [SEC 1: Elliptic Curve Cryptography Version 2.0](https://www.secg.org/sec1-v2.pdf) section 5.1
+ *
+ * This implementaiton is based on https://github.com/ecies/js
+ */
+
+import { chacha20poly1305 } from '@noble/ciphers/chacha';
+import { secp256k1 } from '@noble/curves/secp256k1';
+import { hkdf } from '@noble/hashes/hkdf';
+import { sha256 } from '@noble/hashes/sha256';
+
+/**
+ * Maximum recommended batch size for ECIES encryption.
+ * Limit is based on birthday paradox: with 96-bit random nonces,
+ * collision probability becomes non-negligible after ~2^48 messages.
+ * This conservative limit keeps collision risk below 1 in 10^9.
+ */
+const MAX_BATCH_SIZE = 10000;
+
+/**
+ * ECIES encrypt multiple data items to a public key using a single ephemeral key.
+ * Each message uses the same shared secret but a unique random nonce.
+ * Messages can be decrypted in any order.
+ *
+ * Note: Messages in a batch share an ephemeral key, making them linkable.
+ * For maximum forward secrecy, use eciesEncrypt() for each message individually.
+ *
+ * @param dataArray - Array of data to encrypt (max 10,000 items)
+ * @param publicKeyBytes - 32-byte (Schnorr x-only) or 33-byte (compressed) public key
+ * @returns Array of encrypted messages: [ephemeralPubKey(33) || nonce(12) || ciphertext || tag(16)]
+ * @throws Error if batch size exceeds MAX_BATCH_SIZE
+ */
+export function eciesEncryptBatch(
+  dataArray: Uint8Array[],
+  publicKeyBytes: Uint8Array,
+): Uint8Array[] {
+  if (dataArray.length > MAX_BATCH_SIZE) {
+    throw new Error(
+      `Batch size ${dataArray.length} exceeds maximum ${MAX_BATCH_SIZE}. Split into smaller batches or use eciesEncrypt() for individual messages.`,
+    );
+  }
+
+  // Step 1: Parse and validate the recipient's public key
+  const recipientPublicKey = parsePublicKey(publicKeyBytes);
+
+  // Step 2: Generate single ephemeral key pair for entire batch
+  const ephemeralPrivKey = secp256k1.utils.randomSecretKey();
+  const ephemeralPubKey = secp256k1.Point.BASE.multiply(
+    secp256k1.Point.Fn.fromBytes(ephemeralPrivKey),
+  );
+
+  // Step 3: Compute shared secret once using ECDH
+  const sharedSecret = getSharedSecret(
+    ephemeralPrivKey,
+    recipientPublicKey.toRawBytes(true),
```

Comment:
toRawBytes is also marked as deprecated and it's recommended to replace with just toBytes